prompting/apparmor/message.go

```diff
@@ -0,0 +1,436 @@
+package apparmor
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/snapcore/snapd/arch"
+)
+
+// Message fields are defined as raw sized integer types as the same type may be
+// packed as 16 bit or 32 bit integer, to accommodate other fields in the
+// structure.
+
+// MsgHeader is the header of all apparmor notification messages.
+//
+// The header encodes the size of the entire message, including variable-size
+// components and the version of the notification protocol.
+type MsgHeader struct {
+	// Length is the length of the entire message, including any more
+	// specialized messages appended after the header and any variable-length
+	// objects referenced by any such message.
+	Length uint16
+	// Version is the version of the communication protocol.
+	// Currently version 2 is implemented by the demo kernel based on Linux 5.7.7,
+	// but version 3 was used in the past.
+	Version uint16
+}
+
+// UnmarshalBinary unmarshals the header from binary form.
+func (msg *MsgHeader) UnmarshalBinary(data []byte) error {
+	const prefix = "cannot unmarshal apparmor message header"
+
+	// Unpack fixed-size elements.
+	order := arch.Endian() // ioctl messages are native byte order, but others might not be
+	buf := bytes.NewBuffer(data)
+	if err := binary.Read(buf, order, msg); err != nil {
+		return fmt.Errorf("%s: %s", prefix, err)
+	}
+
+	if msg.Version != 2 {
+		return fmt.Errorf("%s: unsupported version: %d", prefix, msg.Version)
+	}
+	if int(msg.Length) != len(data) {
+		return fmt.Errorf("%s: length mismatch %d != %d",
+			prefix, msg.Length, len(data))
+	}
+
+	return nil
+}
+
+// RequestBuffer returns a new buffer for communication with the kernel.
+// The buffer contains encoded information about its size and protocol version.
+func RequestBuffer() []byte {
+	bufSize := 0xFFFF
+	buf := bytes.NewBuffer(make([]byte, 0, bufSize))
+	header := MsgHeader{Version: 2, Length: uint16(bufSize)}
+	order := arch.Endian() // ioctl requests are native byte order, but others might not be
+	binary.Write(buf, order, &header)
+	buf.Write(make([]byte, bufSize-buf.Len()))
+	return buf.Bytes()
+}
+
+// msgNotificationFilter describes the configuration of kernel-side message filtering.
```

Comment:
(nitpick) maybe cal it `msgNotificationFilterKernel` (or ..Raw) or something to make it more distinct from  MsgNotificationFilter ?

Reply:
I like `msgNotification{Op,Filter,File}Kernel`. Thanks!